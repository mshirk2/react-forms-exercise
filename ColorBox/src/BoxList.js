import React, { useState } from "react";
import { v4 as uuid } from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm"

const BoxList = () => {
    const [boxes, setBoxes] = useState([]);

    const addBox = box => {
        let newBox = {...box, id:uuid()};
        setBoxes(boxes => [...boxes, newBox]);
    };

    const removeBox = id => {
        setBoxes(boxes => boxes.filter(box => box.id !==id));
    }

    const renderBoxes = () => {
        return (
            <div>
                {boxes.map(({id, backgroundColor, width, height}) => 
                    <Box 
                        id={id} 
                        key={id}
                        backgroundColor={backgroundColor}
                        width={width}
                        height={height}
                        handleRemove={removeBox}
                    />
                )}
            </div>
        );
    };

    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
            {renderBoxes()}
        </div>
    );
};

export default BoxList;