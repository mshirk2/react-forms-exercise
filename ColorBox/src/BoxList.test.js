import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

function addBox(boxList, backgroundColor = 'rgb(5, 25, 30)', width = '2', height = '3') {

    fireEvent.change(
        boxList.getByLabelText("Color"), 
        {target: { value: backgroundColor}}
    );
    fireEvent.change(
        boxList.getByLabelText("Width"), 
        {target: { value: width}}
    );
    fireEvent.change(
        boxList.getByLabelText("Height"), 
        {target: { value: height}}
    );

    const submitButton = boxList.getByText("Make Box");
    fireEvent.click(submitButton);
}

it("renders without crashing", function () {
    render(<BoxList />);
});

it("matches snapshot", function () {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("starts with no boxes", function () {
    const boxList = render(<BoxList />);

    expect(boxList.queryByText("X")).toBeFalsy();
});

it("adds new box", function () {
    const boxList = render(<BoxList />);
    addBox(boxList);
    const removeButton = boxList.getByText("X");

    expect(removeButton).toBeTruthy();
    expect(removeButton.previousSibling).toHaveProperty('style.background', 'rgb(5, 25, 30)');
    expect(removeButton.previousSibling).toHaveProperty('style.width', '2em');
    expect(removeButton.previousSibling).toHaveProperty('style.height', '3em');
});

it("clears inputs after adding box", function () {
    const boxList = render(<BoxList />);
    addBox(boxList);
    const removeButton = boxList.getByText("X");

    expect(removeButton).toBeTruthy();
    expect(boxList.getByLabelText("Color").value = "");
    expect(boxList.getByLabelText("Width").value = "");
    expect(boxList.getByLabelText("Height").value = "");
});

it("removes box", function () {
    const boxList = render(<BoxList />);
    addBox(boxList);
    const removeButton = boxList.getByText("X");
    fireEvent.click(removeButton);

    expect(boxList.queryByText("X")).toBeFalsy();
});