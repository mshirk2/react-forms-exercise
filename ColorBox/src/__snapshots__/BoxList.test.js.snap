// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="backgroundColor"
      >
        Color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="Color"
        type="color"
        value=""
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        placeholder="Width"
        type="number"
        value=""
      />
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        placeholder="Height"
        type="number"
        value=""
      />
      <button>
        Make Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
